packages/dynamodb-emulator/client.js

```diff
@@ -0,0 +1,126 @@
+const log = require('logdown')('dynamodb-emulator:client');
+const { getClient } = require('./index');
+
+const waitTimeout = 5 * 1000;
+const requestTimeout = waitTimeout;
+const deamonPath = path.join(__dirname, 'deamon.js');
+const pidPath = path.join(__dirname, '.pid');
+const wait = ms =>
+  new Promise(accept => {
+    setTimeout(accept, ms);
+  });
+const fsExists = file =>
+  new Promise(accept => {
+    fs.exists(file, accept);
+  });
+const get = (unixSocket, urlpath, params = {}) =>
+  new Promise((accept, reject) => {
+    log.info('request', urlpath, params);
+    const start = Date.now();
+    const req = http.request(
+      {
+        socketPath: unixSocket,
+        timeout: requestTimeout,
+        method: 'GET',
+        path: `/${urlpath}?${qs.stringify(params)}`,
+        headers: {
+          'Content-Length': 0,
+        },
+      },
+      res => {
+        const buffers = [];
+        res.once('error', reject);
+        res.on('data', buffer => buffers.push(buffer));
+        res.on('end', () => {
+          log.info('response', urlpath, params, Date.now() - start);
+          accept(JSON.parse(Buffer.concat(buffers)));
+        });
+      },
+    );
+    req.once('timeout', () => reject(new Error('timed out making request')));
+    req.end();
+  });
+
+async function buildEmulatorHandle(unixSocketFile, options) {
+  const start = Date.now();
+  const { url, port, pid } = await get(unixSocketFile, 'launch', options);
+  log.info('got emulator in (ms)', Date.now() - start);
+
+  return {
+    url,
+    port,
+    pid,
+    // for api compat with index.js emulator
+    terminate: async () => {},
+  };
+}
+
+async function requestEmulator(unixSocketFile, options) {
+  const state = await get(unixSocketFile, 'handshake');
+
+  switch (state.status) {
+    case 'success':
+      return buildEmulatorHandle(unixSocketFile, options);
+    case 'version-mismatch':
+      throw new Error('Deamon code has changed... restarting');
+    default:
+      log.error('Invalid state', state);
+      throw new Error(`Unhandled state: ${state.status}`);
+  }
+}
+
+async function startDeamon(hash, unixSocketFile, options) {
+  // file does not exist so we will need to create one.
+  const pidFile = path.join(pidPath, `${hash}.pid`);
+  const fd = fs.openSync(pidFile, 'w');
+  try {
+    // ensure nobody else is writing to this file.
+    flockSync(fd, 'sh');
+  } catch (err) {
+    console.error(err.stack, '<<< failed flock');
+  }
+  const proc = spawn(process.argv0, [deamonPath, hash], {
+    detached: true,
+    env: {
+      ...process.env,
+      NODE_DEBUG: process.env.NODE_DEBUG || '*dynamodb-emulator*',
+    },
+    stdio: [
+      'ignore',
+      fs.openSync(path.join(pidPath, `${hash}-stdout.log`), 'w'),
+      fs.openSync(path.join(pidPath, `${hash}-stderr.log`), 'w'),
+    ],
+  });
+  proc.unref();
+
+  const start = Date.now();
+  while (Date.now() - start < waitTimeout) {
+    if (await fsExists(unixSocketFile)) {
```

Comment:
Curious why you didn't use `fs.existsSync`

Reply:
I didn't want to use a synchronous method in a while true loop . Easy to effectively deadlock node this way.